src/lib.rs

```diff
@@ -1408,6 +1538,356 @@ where
+    /// [RFC 6749](https://tools.ietf.org/html/rfc6749).
+    ///
+    pub fn add_extra_param<N, V>(mut self, name: N, value: V) -> Self
+    where
+        N: Into<Cow<'a, str>>,
+        V: Into<Cow<'a, str>>,
+    {
+        self.extra_params.push((name.into(), value.into()));
+        self
+    }
+
+    ///
+    /// Appends a new scope to the token request.
+    ///
+    pub fn add_scope(mut self, scope: Scope) -> Self {
+        self.scopes.push(Cow::Owned(scope));
+        self
+    }
+
+    fn prepare_request<RE>(self) -> Result<HttpRequest, RequestTokenError<RE, TE>>
+    where
+        RE: Error + 'static,
+    {
+        Ok(endpoint_request(
+            self.auth_type,
+            self.client_id,
+            self.client_secret,
+            &self.extra_params,
+            None,
+            Some(&self.scopes),
+            self.device_authorization_url
+                .ok_or_else(|| {
+                    RequestTokenError::Other("no device authorization_url provided".to_string())
+                })?
+                .url(),
+            vec![],
+        ))
+    }
+
+    ///
+    /// Synchronously sends the request to the authorization server and awaits a response.
+    ///
+    pub fn request<F, RE, EF>(
+        self,
+        http_client: F,
+    ) -> Result<DeviceAuthorizationResponse<EF>, RequestTokenError<RE, TE>>
+    where
+        F: FnOnce(HttpRequest) -> Result<HttpResponse, RE>,
+        RE: Error + 'static,
+        EF: ExtraDeviceAuthorizationFields,
+    {
+        http_client(self.prepare_request()?)
+            .map_err(RequestTokenError::Request)
+            .and_then(endpoint_response)
+    }
+
+    ///
+    /// Asynchronously sends the request to the authorization server and returns a Future.
+    ///
+    pub async fn request_async<C, F, RE, EF>(
+        self,
+        http_client: C,
+    ) -> Result<DeviceAuthorizationResponse<EF>, RequestTokenError<RE, TE>>
+    where
+        C: FnOnce(HttpRequest) -> F,
+        F: Future<Output = Result<HttpResponse, RE>>,
+        RE: Error + 'static,
+        EF: ExtraDeviceAuthorizationFields,
+    {
+        let http_request = self.prepare_request()?;
+        let http_response = http_client(http_request)
+            .await
+            .map_err(RequestTokenError::Request)?;
+        endpoint_response(http_response)
+    }
+}
+
+///
+/// The request for an device access token from the authorization server.
+///
+/// See https://tools.ietf.org/html/rfc8628#section-3.4.
+///
+#[derive(Clone)]
+pub struct DeviceAccessTokenRequest<'a, TR, TT, EF>
+where
+    TR: TokenResponse<TT>,
+    TT: TokenType,
+    EF: ExtraDeviceAuthorizationFields,
+{
+    auth_type: &'a AuthType,
+    client_id: &'a ClientId,
+    client_secret: Option<&'a ClientSecret>,
+    extra_params: Vec<(Cow<'a, str>, Cow<'a, str>)>,
+    token_url: Option<&'a TokenUrl>,
+    dev_auth_resp: &'a DeviceAuthorizationResponse<EF>,
+    time_fn: Arc<dyn Fn() -> DateTime<Utc> + 'a + Send + Sync>,
```

Comment:
I think we can probably get away with a shorter lifetime than `'a` here (i.e., `time_fn` only needs to live as long as the `DeviceAccessTokenRequest`; it doesn't need to live as long as the `Client` object, which is what bounds `'a`). does Rust require an explicit lifetime here? if so, I think we should add a separate `'b`.

Reply:
Looks like it does require a separate lifetime. I'll add one in.